### ISSUE TITLE
Conservative ALE formulation

### ISSUE THREAD

**Issue by tkarna** [thetisproject/thetis#57]
On a static mesh a forward Euler update of field `T` can be expressed as

```
M*T_{n+1} - M*T_{n} = dt*L
```

where `M` is the mass matrix and `L` is the linear form of the equation.
If the mesh is time dependent, the conservative arbitrary Lagrangian-Eulerian (ALE) formulation becomes

```
M_{n+1}*T_{n+1} - M_{n}*T_{n} = dt*L
```

where the mass matrices now depend on time. Here `L` is also modified to take into account the mesh velocity.

What is the easiest way to implement this in firedrake? My guess is that it would make sense to compute  and store `M_{n} T_{n}` in the previous time step and then use it to compute `T_{n+1}` the next iteration. Thus there would be no need to store the mass matrices.

Alternatively you could cast the old solution to the new mesh and then do a conventional solve on a static mesh:

```
M_{n+1}*T_{n+1} - M_{n+1}*P*T_{n} = dt*L
```

with `P = inv(M_{n+1})*M_{n}`. But this is probably costly.

This should work on arbitrary function spaces: The formulation applies to both tracers (which are most likely DG) and velocity (which can be partially continuous).

**Comment by stephankramer** [comment]
Some thoughts:
* Your first suggestion indeed seems easiest
* I guess things get a little more complicated with implicit time integration. In a 2-level method you'd want to evaluate L with mesh position at t^{n+1/2} to maintain 2nd order. For multi-level RK I guess you'd need to move the mesh within each sub-step? This all still based on the linearisation of the mesh movement, whereas you could go one step further and treat the mesh movement as part of the nonlinear system (I'm thinking maybe 3D wetting and drying). Anyway, that's probably not all that relevant for now...
* Note that switching to conservative in the mass terms also requires to change to conservative form in the associated grid velocity term. 

So you do either:

    M^n T^{n+1} - M^n T^n + v^n (u-u_g).grad(T^n) + ...

(i.e. we're basically evaluating the equations in a moving frame of reference), or:

    M^{n+1} T^{n+1} - M^n T^n + v^n div( (u-u_g)*T^n) + ...

where `u_g` is the grid velocity.

**Comment by tkarna** [comment]
Yes, all valid points. I agree that for multilevel RK moving the mesh every sub-step is probably easiest. It should be OK if the mesh update is not very expensive. Alternative is to update only at the last step, but that should be done carefully to ensure conservation and tracer consistency.

My next question is about the implementation. I guess `M_{n} T_{n}` can be evaluated just by assembling it and storing it in a `Function`. It's a bit unclear however how that can be incorporated in the `T_{n+1}` solve however. If one assembles `L` as well, one could then use `dt*L + M_{n} T_{n}` in the solve? Can this be done with firedrake solvers or does it require using PETSc API?

**Comment by wence-** [comment]
I think you replace `M_n T_n` by `action(mass_matrix, T_n)` in your form, no?  Or have a missed something?

**Comment by tkarna** [comment]
I see. I'll give that a shot.

**Comment by stephankramer** [comment]
Maybe I'm missing something, but how is adding action(mass_matrix, T_n) different from adding dot(v_tst, T_n)*dx to the rhs? In both cases you would need T_n to refer to a different copy of the mesh which still stores the old coordinates so that it presumably knows this specific term needs to be evaluated using different coordinates; Is that even allowed having different topologically equivalent meshes with different positions in the same form? 

Or is mass_matrix a pre-assembled `Matrix` here?

I would have thought the way to do this is to assemble this rhs contribution in the previous time step indeed (before the mesh has moved) into a Function and then add this to the rest of the rhs that you assemble in the next timestep. This would require using the linear solver interface that takes a `Matrix`  and `Vector` however which is little ugly and I have no idea how you'd this for the nonlinear case. Also this would presumably break the adjoint.

**Comment by tkarna** [comment]
I think that we'd wouldn't want to have a old copy of the mesh even if it is allowed, due to memory consumption. So one way of doing it would be to assemble and store the old mass matrix, but I don't think it's needed; The latter approach you mention seems legit. I've never implemented such solver though so it's a bit vague to me. I'll give it a shot.

**Comment by wence-** [comment]
@stephankramer yes, I'm not understanding what's going on properly I think.

@tkarna you can have two meshes that share mesh data but have different coordinate fields.  See http://firedrakeproject.org/mesh-coordinates.html#changing-the-coordinate-function-space.  There's also a way to reinterpret fields on the old mesh as fields on the new mesh http://firedrakeproject.org/mesh-coordinates.html#replacing-the-mesh-geometry-of-an-existing-function

**Comment by wence-** [comment]
One thing we could consider (because having different meshes is weird) is to be able to optionally pass a vector-valued function into dx which is the "coordinates at which to evaluate the integral".  @colinjcotter can maybe comment on how he does ALE stuff (especially the L part).  IIRC his verbal explanation he does an Eulerian half step of `L` with `u - v`, then a Lagrangian step with `v` and then an Eulerian half step again with `u - v`.

**Comment by tkarna** [comment]
Thanks @wence- , good to know. I have the impression however that we do not need to store the old mesh or its coordinates.

Following @stephankramer 's comments, the following seems to work for the forward Euler case:

1. given state and mesh at `t_{n}`
1. update 2D equations
1. compute mesh velocity from 2D elevation change and dt
1. solve 3D equations in the old mesh
   - uses conservative advection operator with velocity `w - w_mesh`
1. pre-assemble linear forms for all 3D fields:

    ```python
    sol_lin_form = Function(fs_3d)
    assemble(inner(solution, test)*dx, sol_lin_form)
    ```

1. update mesh coordinates to match new elevation
1. update 3D fields to new geometry:

    ```python
    A = assemble(inner(trial, test)*dx)
    solve(A, solution, sol_lin_form)
    ```

1. Correct 3D velocity to match 2D velocity
1. Update w

Step 7. essentially applies the operator `P = inv(M_{n+1})*M_{n}` I mentioned in the first comment.

This presumably works for nonlinear equations as well because the solver is the same, there's just a correction operation on the new solution. (Assuming that mesh movement is not a part of the nonlinear system - should be fine for current 3D solvers). Generalization to multi-stage RK methods should be possible as well.

Note that here the rhs is evaluated in the old mesh. This gets a bit more complex if it needs to be evaluated in another geometry, say at `t_{n+1/2}` as @stephankramer suggested.

Can you think of a better implementation? Here one might have to allocate one additional function per time stepper (for the linear form), but otherwise memory usage should be similar. The matrix `A` could be shared with the normal explicit solver. Compared to a static mesh, this is obviously quite a bit more costly as `A` needs to be reassembled, but I guess there's no way around it. Maybe we could somehow cache the matrices, e.g. have one per function space.

**Comment by wence-** [comment]
I don't think caching the matrix is the right idea.  Invalidating these things is always messy.  As you note, you only need one mass matrix per function space, so maybe the right interface for step 7 is something like:

```python
def move_fields(fields, rhses):
    solvers = {}
    for f, r in zip(fields, rhses):
        V = f.function_space()
        solver = solvers.setdefault(V, LinearSolver(assemble(inner(test, trial)*dx))))
        solver.solve(f, r)
```

**Comment by stephankramer** [comment]
Hm I'm not sure I quite follow step 4. When you say this handles nonlinear solves as well, do you mean
step 4 is to solving:

    M^n u^* - M^n u^n + \Delta t L(u^*) = 0

on the old mesh, and then step 7:

   M^n+1 u^n+1 = M^n u^*

I'm not sure that would be quite right - in particular in combination with the conservative advection terms. Take for instance the case that u is constant, but w_g is not; in particular that means (u,v,w-w_g) is not divergence free and thus the advection term will create sources and sinks for u which will create artificial gradients in u^* which are then picked up in L.

Note that for fully explicit you don't need the solve in step 4 at all (unless I'm missing something):

    L = ... # some 1-form based on solution_old
    lhs = assemble(inner(test,solution_old)*dx - dt*L)
    # now move mesh
    A = assemble(inner(test, tri)*dx)
    solve(A, solution)

For fully implicit (backward Euler) you could do:

   mass_old = assemble(inner(test, solution_old)*dx))
   # now move mesh
   F = inner(test, tri) * dx + dt*L(test, tri)
   A = assemble(lhs(F))
   L = assemble(rhs(F)) + mass_old
   solver = LinearSolver(A)
   solver.solve(solution, L)

So if you'd do that as sub-steps that would give you DIRK as well I think? That is if everything can be linearized - not sure how you would do nonlinear.

**Comment by tkarna** [comment]
@stephankramer I think you are right, this only works for linear equations. For nonlinear equations you'd need to be able to solve equations where terms are evaluated on different meshes?

> Note that for fully explicit you don't need the solve in step 4 at all 

That's obviously true.

> For fully implicit (backward Euler) you could do:

Yes, I think we could get DIRK and therefore SSP-IMEX schemes as well. As long as the equations are linear, which is the case for the 3D solvers.

**Comment by tkarna** [comment]
@wence- Yes that looks like a good approach. There will normally be two tracer fields so this would save a little.

**Comment by tkarna** [comment]
If I create a linear solver `solver = LinearSolver(A)`, is it possible to update it's matrix `A`? Or should I just create a new solver each time step?

**Comment by colinjcotter** [comment]
I definitely prefer Lawrence's solution, pass a vector field as an argument
to dx to use as a coordinate for that particular term.

cheers
--cjc

On 14 July 2016 at 19:31, Tuomas Kärnä <notifications@github.com> wrote:

> If I create a linear solver solver = LinearSolver(A), is it possible to
> update it's matrix A? Or should I just create a new solver each time step?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/thetisproject/thetis/issues/57#issuecomment-232752014>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AF0q8pDXAFh9N76bvW_pMgJvIcigXmraks5qVoB8gaJpZM4JK5p2>
> .
>



-- 
http://www.imperial.ac.uk/people/colin.cotter

www.cambridge.org/9781107663916

**Comment by tkarna** [comment]
That would be a nice feature. We could then conceivably implement generic ALE time integrators.
Do you think there's any performance penalty involved, i.e. is doing
 `L = term1*dx(coordinates=coords1) + term2*dx(coordinates=coords2)`
more expensive compared to evaluating the terms on the current mesh?

**Comment by wence-** [comment]
So what will happen in the short term is that will be turned into a kernel that implements term1 (taking coords1 as the coordinates) and a kernel implementing term2 (taking coords2 as the coordinates).  These will execute as separate mesh loops (but the upcoming loop fusion stuff in pyop2 will trivially fix this).  If you evaluated the terms separately on separate meshes, then you'd have case (a) anyway.

**Comment by wence-** [comment]
The problem with this approach is how to get the information from about the coordinates into the assembler.  Let me elaborate:

```python
m = UnitIntervalMesh(1)
ca = Function(m.coordinates)
cb = Function(m.coordinates)
cb *= 2

a = term1*dx(metadata={'coordinates': ca}) + term2*dx(metadata={'coordinates': cb})
```

Now, tsfc will generate two kernels for this form, because UFL splits integrals based (in part) on the metadata.  However, I don't see a good way of sending the information back to firedrake about which coordinate field matches which kernel.  I had a go with multiple meshes `term1*dx(domain=m1) + term2*dx(domain=m2)`.  This works but it's a bit hacky.  You effectively only use the second mesh as the domain argument:

```python
from firedrake import *
m = UnitIntervalMesh(1)
f = Function(m.coordinates)
f *= 2
m2 = Mesh(f)
V = FunctionSpace(m, 'DG', 0)
u = TrialFunction(V)
v = TestFunction(V)
a = u*v*dx(domain=m) + u*v*dx(domain=m2)
assemble(m).M.values
=> array([[ 3.]])
```

But this breaks if I write `assemble(u*v*dx(domain=m2))` because the domains of the form are determined from the measure domain and the coefficient domains.  But this form has no coefficients, but the domains of the arguments do not match the domain of the measure.

Ugh.

Thoughts?

**Comment by wence-** [comment]
If you use the firedrake branch `multiple-domains` and the tsfc branch `multiple-domains` you can try this `dx(domain=m1) + dx(domain=m2)` approach out.  Note the caveats in the commit message in firedrake.  If this is properly useful, we can consider proposing some fixes in UFL.  Alternately, we could consider having a special way of doing ALE that requires us to specify the coordinate field only.  This will also require some changes in UFL to pass this information around correctly.

**Comment by tkarna** [comment]
Thanks @wence- . Update: I'm having some issues with the `multiple-domains` branches, documented in the commit firedrakeproject/firedrake@8c6d3a61b51. Meanwhile I have a drafted an ALE time integrator with the first approach. It seems to work, but the implementation would be cleaner with if multiple domains were supported.

**Comment by wence-** [comment]
I think the problem you saw has now been fixed in latest UFL.  Please try again.

**Comment by wence-** [comment]
I have merged these branches that support this approach.  It still fails when integrating an integrand of purely one domain and specifying the "other" mesh in the domain argument to the measure.  But it's a start.  Changing that purely requires us to get UFL working right.

**Comment by tkarna** [comment]
ALE time integrators are now addressed in #70 #73 #74, hence closing.
